src/test/scala/query/MergeQuerySpec.scala

```diff
@@ -0,0 +1,159 @@
+      create.toQuery shouldBe """MATCH (a1:label1)--(a2:label2) MERGE (a1)-[:label]-(a2) RETURN DISTINCT a1"""
+    }
+
+    "add identifier to reference if it appears in the return statement" in {
+      val create = matchPath merge createPath returnDistinct createRelationship
+      val query = create.toQuery
+      query should include (")-[a1:label]-(")
+      query should endWith ("RETURN DISTINCT a1")
+    }
+
+    "allow multiple match paths and" must {
+
+      val otherNode = AnyNode()
+      val matchPath2 = startNode --> otherNode
+      val createPath = startNode -- createRelationship -- endNode --> otherNode
+
+      "assign identifiers to the necessary elements" in {
+        val merge = MergeQuery(createPath, Seq(matchPath, matchPath2))
+        val query = merge.toQuery
+        query should startWith regex """MATCH \(a\d:label1\)--\(a\d:label2\), \(a\d:label1\)-->\(a\d\)"""
+        query should endWith regex """MERGE \(a\d\)-\[:label]-\(a\d\)-->\(a\d\)"""
+      }
+
+      "assign identifiers appropriately" when {
+
+        val merge = MergeQuery(createPath, Seq(matchPath, matchPath2))
+          .onCreate(Seq(startNode.property("a").set("b")))
+          .onMerge(Seq(createRelationship.property("c").set("d"), otherNode.property("e").set("f")))
+
+        def extractIds(query: String) = {
+          val Identifiers = """MATCH \((\w+):label1\)--\((\w+):label2\), \((\w+):label1\)-->\((\w+)\).*""".r
+          val CreateIdentifier = """.*\[(\w*):label\].*""".r
+
+          query match {
+            case Identifiers(startId, endId, startId2, otherId) if startId == startId2 =>
+              query match {
+                case CreateIdentifier(relationshipId) => (startId, endId, otherId, relationshipId)
+              }
+          }
+        }
+
+        "created node/relationship is not referenced" in {
+          val query = (merge onMerge Seq()).toQuery
+          val (startId, endId, otherId, relationshipId) = extractIds(query)
+
+          val mergeTest = s"""MERGE ($startId)-[$relationshipId:label]-($endId)-->($otherId)"""
+          val onCreateTest = s"""ON CREATE SET $startId.a = "b""""
+          query should include (mergeTest)
+          query should include (onCreateTest)
+          query should not include ("""ON MATCH SET""")
+          relationshipId shouldBe ""
+        }
+
+        "multiple properties are set" in {
+          val query = merge.toQuery
+          val (startId, endId, otherId, relationshipId) = extractIds(query)
+
+          val mergeTest = s"""MERGE ($startId)-[$relationshipId:label]-($endId)-->($otherId)"""
+          val onCreateTest = s"""ON CREATE SET $startId.a = "b""""
+          val onMergeTest = s"""ON MATCH SET $relationshipId.c = "d", $otherId.e = "f""""
+          query should include (onMergeTest)
+        }
+
+        "multiple properties are set and a return clause is given" in {
+          val query = (merge returns createRelationship).toQuery
```

Comment:
the queries themselves are different though, i don't recall the nicer way of dealing with this in scalatest..